File: app/src/test/java/com/example/david/lists/view/userlistlist/UserListListLogicTest.java
package com.example.david.lists.view.userlistlist;

import com.example.david.lists.SchedulerProviderMockInit;
import com.example.david.lists.data.datamodel.UserList;
import com.example.david.lists.data.repository.IRepositoryContract;
import com.example.david.lists.data.repository.IRepositoryContract.Repository;
import com.example.david.lists.util.ISchedulerProviderContract;
import com.example.david.lists.util.IUtilNightModeContract;
import com.example.david.lists.view.authentication.IAuthContract;
import com.example.david.lists.view.authentication.IAuthContract.AuthResult;
import com.example.david.lists.view.userlistlist.IUserListViewContract.Adapter;
import com.example.david.lists.view.userlistlist.IUserListViewContract.View;
import com.example.david.lists.view.userlistlist.IUserListViewContract.ViewModel;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Spy;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.disposables.CompositeDisposable;

import static org.hamcrest.Matchers.contains;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class UserListListLogicTest {

    @Mock
    private IUserListViewContract.View view;

    @Mock
    private ViewModel viewModel;

    @Mock
    private Repository repo;

    @Mock
    private IRepositoryContract.UserRepository userRepo;

    @Mock
    private ISchedulerProviderContract schedulerProvider;

    @Spy
    private CompositeDisposable disposable;

    @Mock
    private IUtilNightModeContract utilNightMode;


    @Mock
    private Adapter adapter;


    @InjectMocks
    private UserListListLogic logic;


    private String id = "id";
    private String title = "title";
    private int position = 0;

    private UserList userList = new UserList(id, new UserList(title, position));
    private UserList userListTwo = new UserList(id + 2, new UserList(title + 2, position + 1));


    private List<UserList> userListList = new ArrayList<>(Collections.singletonList(userList));

    private int invalidPosition = -1;
    private String errorMsg = "error";


    @Before
    public void setUp() {
        SchedulerProviderMockInit.INSTANCE.init(schedulerProvider);
    }


    /**
     * Normal behavior
     * - Set View state loading.
     * - Get List from repo.
     * - Save List to ViewModel.
     * - Submit the List to the View.
     * - Set View state display list.
     */
    @Test
    public void onStart() {
        when(repo.getAllUserLists()).thenReturn(Flowable.just(userListList));
        when(viewModel.getViewData()).thenReturn(userListList);

        logic.onStart();

        verify(view).setStateLoading();
        verify(viewModel).setViewData(userListList);
        verify(view).submitList(userListList);
        verify(view).setStateDisplayList();
    }

    /**
     * Error behavior - empty List returned by repo,
     * - Set View state loading.
     * - Get List from repo.
     * - Save List to ViewModel.
     * - Submit the List to the View.
     * - Set View state error with message from the ViewModel.
     */
    @Test
    public void onStartEmptyList() {
        List<UserList> emptyList = new ArrayList<>();

        when(repo.getAllUserLists()).thenReturn(Flowable.just(emptyList));
        when(viewModel.getViewData()).thenReturn(emptyList);
        when(viewModel.getErrorMsgEmptyList()).thenReturn(errorMsg);

        logic.onStart();

        verify(view).setStateLoading();
        verify(viewModel).setViewData(emptyList);
        verify(view).submitList(emptyList);
        verify(view).setStateError(errorMsg);
    }

    /**
     * Error behavior - the repo throws an error,
     * - Set View state loading.
     * - Get List from repo.
     * - Repo throws an error.
     * - Set View state error with message from ViewModel.
     */
    @Test
    public void onStartRepoThrowsError() {
        Throwable throwable = new Throwable();

        when(repo.getAllUserLists()).thenReturn(Flowable.error(throwable));
        when(viewModel.getErrorMsg()).thenReturn(errorMsg);

        logic.onStart();

        verify(view).setStateLoading();
        verify(view).setStateError(errorMsg);
    }


    /**
     * Normal behavior - valid position,
     * - Get the selected UserList with the position argument.
     * - Pass that UserList to {@link View#openUserList(UserList)}
     */
    @Test
    public void userListSelectedValidPosition() {
        when(viewModel.getViewData()).thenReturn(userListList);

        logic.userListSelected(position);

        verify(view).openUserList(userList);
    }

    /**
     * Error behavior - invalid position,
     * - Throws an exception.
     */
    @Test(expected = ArrayIndexOutOfBoundsException.class)
    public void userListSelectedInvalidPosition() {
        when(viewModel.getViewData()).thenReturn(userListList);

        logic.userListSelected(invalidPosition);
    }


    /**
     * Normal behavior
     * - {@link View#openAddDialog(int)} is invoked with the current size
     * of List containing the view data.
     */
    @Test
    public void add() {
        int size = userListList.size();

        when(viewModel.getViewData()).thenReturn(userListList);

        logic.add();

        verify(view).openAddDialog(size);
    }


    /**
     * Normal behavior
     * - Get the UserList at the passed-in position.
     * - Invoked {@link View#openEditDialog(UserList)}.
     */
    @Test
    public void edit() {
        int editPosition = 0;
        List<UserList> editUserLists = new ArrayList<>();
        editUserLists.add(userList);

        when(viewModel.getViewData()).thenReturn(editUserLists);

        logic.edit(editPosition);

        verify(view).openEditDialog(userList);
    }

    /**
     * Error behavior - invalid position,
     * - Exception is thrown.
     */
    @Test(expected = IllegalArgumentException.class)
    public void editInvalidPosition() {
        logic.edit(invalidPosition);
    }


    /**
     * Normal behavior
     * - Updated the Adapter with {@link Adapter#move(int, int)}
     * - Update the ViewModel's view data.
     */
    @Test
    public void dragging() {
        int fromPos = 0;
        int toPos = 1;

        List<UserList> userListsDragging = new ArrayList<>();
        userListsDragging.add(userList);
        userListsDragging.add(userListTwo);

        when(viewModel.getViewData()).thenReturn(userListsDragging);

        logic.dragging(fromPos, toPos, adapter);

        verify(adapter).move(fromPos, toPos);
        assertThat(userListsDragging, is(contains(userListTwo, userList)));
    }


    /**
     * Normal behavior
     * - Get the moved UserList from the ViewModel.
     * - Invoke {@link Repository#updateUserListPosition(UserList, int, int)}.
     */
    @Test
    public void movedPermanently() {
        int newPosition = 0;
        List<UserList> moveUserLists = new ArrayList<>();
        moveUserLists.add(userList);

        when(viewModel.getViewData()).thenReturn(moveUserLists);

        logic.movedPermanently(newPosition);

        verify(repo).updateUserListPosition(userList, userList.getPosition(), position);
    }

    /**
     * Error behavior - invalid position,
     * - Throws an Exception
     */
    @Test(expected = IllegalArgumentException.class)
    public void movedPermanentlyInvalidPosition() {
        logic.movedPermanently(invalidPosition);
    }


    /**
     * Normal behavior
     * - Remove from the Adapter.
     * - Save the deleted UserList itself to the ViewModel's temp list.
     * - Save the deleted UserList's position to the ViewModel.
     * - Remove from the view data in the ViewModel.
     * - Notify user of deletion.
     */
    @Test
    public void delete() {
        List<UserList> tempUserLists = new ArrayList<>();

        when(viewModel.getViewData()).thenReturn(userListList);
        when(viewModel.getTempList()).thenReturn(tempUserLists);
        when(viewModel.getMsgDeletion()).thenReturn(errorMsg);

        logic.delete(position, adapter);

        assertThat(tempUserLists, is(contains(userList)));
        assertThat(userListList.isEmpty(), is(true));
        verify(adapter).remove(position);
        verify(viewModel).setTempPosition(position);
        verify(view).notifyUserOfDeletion(errorMsg);
    }

    /**
     * Error behavior - invalid position,
     * - Exception is thrown
     */
    @Test(expected = IllegalArgumentException.class)
    public void deleteInvalidPosition() {
        logic.delete(invalidPosition, adapter);
    }


    /**
     * Normal behavior
     * - Re-add to the adapter.
     * - Re-add to the view data List.
     * - Remove the last UserList from the temp List.
     * - Because the temp List is not empty, pass the temp List to the repo for deletion.
     * - Clear the temp List.
     */
    @Test
    public void undoRecentDeletion() {
        int tempPosition = 0;

        List<UserList> tempList = new ArrayList<>();
        tempList.add(userList);
        tempList.add(userListTwo);
        List<UserList> viewDataList = new ArrayList<>();

        when(viewModel.getTempList()).thenReturn(tempList);
        when(viewModel.getTempPosition()).thenReturn(tempPosition);
        when(viewModel.getViewData()).thenReturn(viewDataList);

        logic.undoRecentDeletion(adapter);

        verify(adapter).reAdd(tempPosition, userListTwo);
        verify(repo).deleteUserLists(tempList);
        assertThat(viewDataList, is(contains(userListTwo)));
        assertThat(tempList.isEmpty(), is(true));
    }

    /**
     * Error behavior - empty temp List,
     * - An exception is thrown
     */
    @Test(expected = UnsupportedOperationException.class)
    public void undoRecentDeletionEmptyList() {
        when(viewModel.getTempList()).thenReturn(new ArrayList<>());

        logic.undoRecentDeletion(adapter);
    }

    /**
     * Error behavior - invalid position,
     * - An exception is thrown
     */
    @Test(expected = UnsupportedOperationException.class)
    public void undoRecentDeletionInvalidPosition() {
        when(viewModel.getTempList()).thenReturn(userListList);
        when(viewModel.getTempPosition()).thenReturn(invalidPosition);

        logic.undoRecentDeletion(adapter);
    }


    /**
     * Normal behavior
     * - Check if the temp List is empty.
     * - Pass the temp List {@link Repository#deleteUserLists(List)}.
     * - Clear the temp List.
     */
    @Test
    public void deletionNotificationTimedOut() {
        when(viewModel.getTempList()).thenReturn(userListList);

        logic.deletionNotificationTimedOut();

        verify(repo).deleteUserLists(userListList);
        assertThat(userListList.isEmpty(), is(true));
    }

    /**
     * Error behavior - empty temp List,
     * - Check if the temp List is empty.
     * - Return without invoking any other method.
     */
    @Test
    public void deletionNotificationTimedOutEmptyTempList() {
        when(viewModel.getTempList()).thenReturn(new ArrayList<>());

        logic.deletionNotificationTimedOut();

        verify(repo, never()).deleteUserLists(anyList());
    }


    /**
     * Normal behavior
     * - Invokes {@link View#confirmSignOut()}
     */
    @Test
    public void signOut() {
        logic.signOut();

        verify(view).confirmSignOut();
    }


    /**
     * Normal behavior
     * - Invoke {@link View#openAuthentication(IAuthContract.AuthGoal, int, String)}
     */
    @Test
    public void signOutConfirmed() {
        IAuthContract.AuthGoal authGoal = IAuthContract.AuthGoal.SIGN_OUT;
        int requestCode = 100;
        String intentExtraKey = "key";

        when(viewModel.getRequestCode()).thenReturn(requestCode);
        when(viewModel.getIntentExtraAuthResultKey()).thenReturn(intentExtraKey);

        logic.signOutConfirmed();

        verify(view).openAuthentication(authGoal, requestCode, intentExtraKey);
    }


    /**
     * Normal behavior
     * - Invoke {@link View#openAuthentication(IAuthContract.AuthGoal, int, String)}
     */
    @Test
    public void signIn() {
        IAuthContract.AuthGoal authGoal = IAuthContract.AuthGoal.SIGN_IN;
        int requestCode = 100;
        String intentExtraKey = "key";

        when(viewModel.getRequestCode()).thenReturn(requestCode);
        when(viewModel.getIntentExtraAuthResultKey()).thenReturn(intentExtraKey);

        logic.signIn();

        verify(view).openAuthentication(authGoal, requestCode, intentExtraKey);
    }


    /**
     * Normal behavior
     * - {@link AuthResult#AUTH_SUCCESS} parameter
     * - Invoke {@link View#recreateView()}
     */
    @Test
    public void authResult() {
        logic.authResult(AuthResult.AUTH_SUCCESS);

        verify(view).recreateView();
    }

    /**
     * Error behavior
     * - {@link AuthResult#AUTH_FAILED} parameter
     * - Nothing invoked.
     */
    @Test
    public void authResultFailed() {
        logic.authResult(AuthResult.AUTH_FAILED);

        verify(view, never()).recreateView();
    }

    /**
     * Error behavior
     * - {@link AuthResult#AUTH_CANCELLED} parameter
     * - Nothing invoked.
     */
    @Test
    public void authResultCancelled() {
        logic.authResult(AuthResult.AUTH_CANCELLED);

        verify(view, never()).recreateView();
    }


    /**
     * Normal behavior
     * - If the menu item is unchecked, Night mode should be enabled.
     */
    @Test
    public void setNightModeUnchecked() {
        logic.setNightMode(false);

        verify(utilNightMode).setNight();
    }

    /**
     * Normal behavior
     * - If the menu item is checked, Day mode should be enabled.
     */
    @Test
    public void setNightModeChecked() {
        logic.setNightMode(true);

        verify(utilNightMode).setDay();
    }


    /**
     * Normal behavior
     * - Verify true is returned when the user is anonymous.
     */
    @Test
    public void isUserAnon() {
        boolean isAnon = true;

        when(userRepo.isAnonymous()).thenReturn(isAnon);

        assertThat(
                logic.isUserAnon(),
                is(isAnon)
        );
    }

    /**
     * Normal behavior
     * - Verify true is returned when night mode is enabled.
     */
    @Test
    public void isNightModeEnabled() {
        boolean nightModeState = true;

        when(utilNightMode.isNightModeEnabled()).thenReturn(nightModeState);

        assertThat(
                logic.isNightModeEnabled(),
                is(nightModeState)
        );
    }


    /**
     * Normal behavior
     * - Verify {@link CompositeDisposable#clear()} is invoked.
     */
    @Test
    public void onDestroy() {
        logic.onDestroy();
        verify(disposable).clear();
    }
}

File: app/src/test/java/com/example/david/lists/view/userlistlist/UserListListLogicTest.kt
package com.example.david.lists.view.userlistlist

import com.example.david.lists.SchedulerProviderMockInit
import com.example.david.lists.data.datamodel.UserList
import com.example.david.lists.data.repository.IRepositoryContract
import com.example.david.lists.util.ISchedulerProviderContract
import com.example.david.lists.util.IUtilNightModeContract
import io.mockk.*
import io.reactivex.Flowable
import io.reactivex.disposables.CompositeDisposable
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource

class UserListListLogicTest {

    private val view = mockk<IUserListViewContract.View>(relaxUnitFun = true)

    private val viewModel = mockk<IUserListViewContract.ViewModel>(relaxUnitFun = true)

    private val repo = mockk<IRepositoryContract.Repository>(relaxUnitFun = true)

    private val schedulerProvider = mockk<ISchedulerProviderContract>()

    private val disposable = spyk<CompositeDisposable>()

    private val utilNightMode = mockk<IUtilNightModeContract>(relaxUnitFun = true)

    private val logic = UserListListLogic(view, viewModel, utilNightMode, repo, schedulerProvider, disposable)


    private val adapter = mockk<IUserListViewContract.Adapter>(relaxUnitFun = true)


    private val id = "id"
    private val title = "title"
    private val position = 0
    private val userListOne = UserList(title, position, id)
    private val userListTwo = UserList(title + 2, position + 1, id + 2)

    private val message = "message"

    private val emptyList = mutableListOf<UserList>()


    @BeforeEach
    fun init() {
        clearAllMocks()
        SchedulerProviderMockInit.init(schedulerProvider)
    }


    @Nested
    inner class OnStart {
        /**
         * - Set the View's data with data from the ViewModel - because the ViewModel's view data is not empty.
         * - Get List from repo.
         * - Save List to ViewModel.
         * - Set the View's data
         * - Set View state display list.
         */
        @Test
        fun `onStart - ViewModel has View Data`() {
            val userListList = mutableListOf(userListOne)

            every { repo.getUserLists } returns Flowable.just(userListList)
            every { viewModel.viewData } returns userListList

            logic.onStart()

            verify(exactly = 2) { view.setViewData(userListList) }
            verify { viewModel.viewData = userListList }
            verify { view.setStateDisplayList() }
        }

        /**
         * - Set View state loading.
         *   - Because the ViewModel's ViewData is empty.
         * - Get List from repo.
         * - Save List to ViewModel.
         * - Submit the List to the View.
         * - Set View state error with message from the ViewModel.
         */
        @Test
        fun `onStart - Empty List from repo`() {
            every { repo.getUserLists } returns Flowable.just(emptyList)
            every { viewModel.viewData } returns emptyList
            every { viewModel.errorMsgEmptyList } returns message

            logic.onStart()

            verify { view.setStateLoading() }
            verify { viewModel.viewData = emptyList }
            verify { view.setViewData(emptyList) }
            verify { view.setStateError(message) }
        }

        /**
         * - Set View state loading.
         *   - Because the ViewModel's ViewData is empty.
         * - Get List from repo.
         * - Repo throws an error.
         * - Set View state error with message from ViewModel.
         */
        @Test
        fun `onStart - Repo throws an error`() {
            val throwable = Throwable()

            every { viewModel.viewData } returns emptyList
            every { repo.getUserLists } returns Flowable.error(throwable)
            every { viewModel.errorMsg } returns message

            logic.onStart()

            verify { view.setStateLoading() }
            verify { view.setStateError(message) }
        }
    }


    @Nested
    inner class UserListSelected {
        /**
         * - Get the selected UserList with the position argument.
         * - Pass that UserList to the View.
         */
        @Test
        fun `userListSelected - Valid Position`() {
            val position = 0
            val userListList = mutableListOf(userListOne)

            every { viewModel.viewData } returns userListList

            logic.userListSelected(position)

            verify { view.openUserList(userListList[position]) }
        }

        /**
         * - Attempt to get the selected UserList from the ViewModel with the invalid position.
         * - Exception is thrown.
         */
        @Test
        fun `userListSelected - Invalid Position`() {
            val invalidPosition = -1
            val userListList = mutableListOf(userListOne)

            every { viewModel.viewData } returns userListList

            assertThrows<IndexOutOfBoundsException> {
                logic.userListSelected(invalidPosition)
            }
        }
    }


    /**
     * - [IUserListViewContract.View.openAddDialog] is invoked with the current size.
     * of List containing the view data.
     */
    @Test
    fun add() {
        val userListList = mutableListOf(userListOne)
        val size = userListList.size

        every { viewModel.viewData } returns userListList

        logic.add()

        verify { view.openAddDialog(size) }
    }


    @Nested
    inner class Edit {
        /**
         * - Get the UserList at the passed-in position.
         * - Pass that UserList to [IUserListViewContract.View.openEditDialog].
         */
        @Test
        fun edit() {
            val position = 0
            val userListList = mutableListOf(userListOne)

            every { viewModel.viewData } returns userListList

            logic.edit(position)

            verify { view.openEditDialog(userListOne) }
        }

        /**
         * - Attempt to get the selected UserList from the ViewModel with the invalid position.
         * - Exception is thrown.
         */
        @Test
        fun `edit - Invalid Position`() {
            val invalidPosition = -1
            val userListList = mutableListOf(userListOne)

            every { viewModel.viewData } returns userListList

            assertThrows<IndexOutOfBoundsException> {
                logic.edit(invalidPosition)
            }
        }
    }


    @Nested
    inner class Dragging {
        /**
         * - Updated the Adapter with [IUserListViewContract.Adapter.move].
         * - Update the ViewModel's view data - verify that the order has changed.
         */
        @Test
        fun dragging() {
            val fromPosition = 0
            val toPosition = 1
            val userListList = mutableListOf(userListOne, userListTwo)

            every { viewModel.viewData } returns userListList

            logic.dragging(fromPosition, toPosition, adapter)

            verify { adapter.move(fromPosition, toPosition) }
            // Assert that the ViewData was updated.
            assertThat(userListList).containsExactly(userListTwo, userListOne)
        }

        /**
         * - Attempt to re-arrange the ViewData with the invalid positions.
         * - Exception is thrown.
         */
        @ParameterizedTest
        @CsvSource("-1, 0", "0, -1", "-1, -1")
        fun `dragging - Invalid positions`(fromPosition: Int, toPosition: Int) {
            val userListList = mutableListOf(userListOne, userListTwo)

            every { viewModel.viewData } returns userListList

            assertThrows<IndexOutOfBoundsException> {
                logic.dragging(fromPosition, toPosition, adapter)
            }
        }
    }


    @Nested
    inner class MovePermanently {
        /**
         * - Get the moved UserList from the ViewModel.
         * - Invoke [IRepositoryContract.Repository.updateUserListPosition].
         */
        @Test
        fun movedPermanently() {
            val newPosition = 0
            val userListList = mutableListOf(userListOne)

            every { viewModel.viewData } returns userListList

            logic.movedPermanently(newPosition)

            verify { repo.updateUserListPosition(userListOne, userListOne.position, newPosition) }
        }

        /**
         * - Attempt to get the selected UserList from the ViewModel with the invalid position.
         * - Exception is thrown.
         */
        @Test
        fun `movedPermanently - InvalidPosition`() {
            val invalidPosition = -1
            val userListList = mutableListOf(userListOne)

            every { viewModel.viewData } returns userListList

            assertThrows<IndexOutOfBoundsException> {
                logic.movedPermanently(invalidPosition)
            }
        }
    }


    @Nested
    inner class Delete {
        /**
         * - Get the deleted UserList from the ViewModel.
         * - Remove from the Adapter.
         * - Save the deleted UserList itself to the ViewModel's temp list.
         * - Save the deleted UserList's position to the ViewModel.
         * - Remove from the view data in the ViewModel.
         * - Notify user of deletion.
         */
        @Test
        fun delete() {
            val position = 0
            val userListList = mutableListOf(userListOne)
            val tempUserLists = mutableListOf<UserList>()

            every { viewModel.viewData } returns userListList
            every { viewModel.tempList } returns tempUserLists
            every { viewModel.msgDeletion } returns message

            logic.delete(position, adapter)

            verify { adapter.remove(position) }
            assertThat(tempUserLists).containsExactly(userListOne)
            verify { viewModel.tempPosition = position }
            assertThat(userListList.isEmpty()).isTrue()
            verify { view.notifyUserOfDeletion(message) }
        }

        /**
         * - Attempt to get the selected UserList from the ViewModel with the invalid position.
         * - Exception is thrown
         */
        @Test
        fun `delete - Invalid Position`() {
            val invalidPosition = -1
            val userListList = mutableListOf(userListOne)
            val tempUserLists = mutableListOf<UserList>()

            every { viewModel.viewData } returns userListList
            every { viewModel.tempList } returns tempUserLists

            assertThrows<java.lang.IndexOutOfBoundsException> {
                logic.delete(invalidPosition, adapter)
            }
        }
    }


    @Nested
    inner class UndoDeletion {
        /**
         * - Get the last added UserList from the ViewModel's temp list.
         * - Re-add to the adapter using the temp position from the ViewModel.
         * - Re-add to the ViewData List.
         * - Remove the last UserList from the temp List.
         * - Because the temp List is not empty, pass the temp List to the repo for deletion.
         * - Clear the temp List.
         */
        @Test
        fun undoRecentDeletion() {
            val userListList = mutableListOf<UserList>()
            val tempUserLists = mutableListOf(userListOne, userListTwo)
            val tempPosition = 0

            every { viewModel.viewData } returns userListList
            every { viewModel.tempList } returns tempUserLists
            every { viewModel.tempPosition } returns tempPosition

            logic.undoRecentDeletion(adapter)

            verify { adapter.reAdd(tempPosition, userListTwo) }
            assertThat(userListList).containsExactly(userListTwo)
            verify { repo.deleteUserLists(tempUserLists) }
            assertThat(tempUserLists.isEmpty()).isTrue()
        }

        /**
         * - The temp List in the ViewModel is empty.
         * - Exception is thrown
         */
        @Test
        fun `undoRecentDeletion - Empty Temp List`() {
            every { viewModel.tempList } returns emptyList
            every { viewModel.errorMsgInvalidUndo } returns message

            assertThrows<UnsupportedOperationException> {
                logic.undoRecentDeletion(adapter)
            }
        }

        /**
         * - The temp position in the ViewModel is invalid.
         * - Exception is thrown
         */
        @Test
        fun `undoRecentDeletion - Invalid Temp Position`() {
            val invalidPosition = -1
            val tempUserLists = mutableListOf(userListOne)

            every { viewModel.tempList } returns tempUserLists
            every { viewModel.tempPosition } returns invalidPosition
            every { viewModel.errorMsgInvalidUndo } returns message

            assertThrows<UnsupportedOperationException> {
                logic.undoRecentDeletion(adapter)
            }
        }
    }


    @Nested
    inner class DeleteNotificationTimeOut {
        /**
         * - Check if the temp List is empty.
         *   - Will not be in this test.
         * - Pass the temp List [IRepositoryContract.Repository.deleteUserLists].
         * - Clear the temp List.
         */
        @Test
        fun deletionNotificationTimedOut() {
            val userListList = mutableListOf(userListOne)

            every { viewModel.tempList } returns userListList

            logic.deletionNotificationTimedOut()

            verify { repo.deleteUserLists(userListList) }
            assertThat(userListList.isEmpty()).isTrue()
        }

        /**
         * - Check if the temp List is empty.
         * - Return without invoking any other method.
         */
        @Test
        fun `deletionNotificationTimedOut - Empty Temp List`() {
            every { viewModel.tempList } returns emptyList

            logic.deletionNotificationTimedOut()

            verify { repo wasNot Called }
        }
    }


    /**
     * - Invokes [IUserListViewContract.View.openPreferences]
     */
    @Test
    fun preferencesSelected() {
        logic.preferencesSelected()

        verify { view.openPreferences() }
    }


    @Nested
    inner class NightMode {
        /**
         * - Menu item is unchecked.
         * - Night mode is enabled.
         */
        @Test
        fun `setNightMode - Unchecked`() {
            logic.setNightMode(false)

            verify { utilNightMode.setNight() }
        }

        /**
         * - Menu item is Checked.
         * - Night mode is Disabled.
         */
        @Test
        fun `setNightMode - Checked`() {
            logic.setNightMode(true)

            verify { utilNightMode.setDay() }
        }

        /**
         * - True is returned when night mode is enabled.
         */
        @Test
        fun isNightModeEnabled() {
            val nightModeEnabled = true

            every { utilNightMode.nightModeEnabled } returns nightModeEnabled

            assertThat(logic.isNightModeEnabled).isEqualTo(nightModeEnabled)
        }
    }


    /**
     * - Verify that [CompositeDisposable.clear] is invoked.
     */
    @Test
    fun onDestroy() {
        logic.onDestroy()

        verify { disposable.clear() }
    }
}